### PR TITLE
Adding policy for CPU Limits for all container types

### PR DESCRIPTION
**Pull request by nsagark**
## Checklist

- [x] I have read the [policy contribution guidelines](https://github.com/kyverno/policies/blob/main/README.md#contribution).
- [x] I have added test manifests and resources covering both positive and negative tests that prove this policy works as intended.
- [x] I have added the artifacthub-pkg.yml file and have verified it is complete and correct.

### CODE REVIEW

File: other/require-cpu-limits/require-cpu-limits.yaml [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    kyverno.io/kubernetes-version: "1.26"
+    policies.kyverno.io/description: >-
+      Setting CPU limits on pods ensures fair distribution of resources, preventing any single pod from monopolizing CPU and impacting the performance of other pods. This practice enhances stability, predictability, and cost control, while also mitigating the noisy neighbor problem and facilitating efficient scaling in Kubernetes clusters. This policy ensures that cpu limits are set on every container.
+spec:
+  validationFailureAction: Audit
+  background: true
+  rules:
+  - name: check-cpu-limits
+    match:
+      resources:
```
Comment (chipzoller) [inline other/require-cpu-limits/require-cpu-limits.yaml:19]: All policies should be using `any`/`all` under the match and exclude conditions.

File: other/require-cpu-limits/require-cpu-limits.yaml [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+metadata:
+  name: require-cpu-limits
+  annotations:
+    policies.kyverno.io/title: Require CPU Limits
+    policies.kyverno.io/category: Other
+    policies.kyverno.io/severity: medium
+    policies.kyverno.io/subject: Pod
+    kyverno.io/kubernetes-version: "1.26"
+    policies.kyverno.io/description: >-
+      Setting CPU limits on pods ensures fair distribution of resources, preventing any single pod from monopolizing CPU and impacting the performance of other pods. This practice enhances stability, predictability, and cost control, while also mitigating the noisy neighbor problem and facilitating efficient scaling in Kubernetes clusters. This policy ensures that cpu limits are set on every container.
```
Comment (chipzoller) [inline other/require-cpu-limits/require-cpu-limits.yaml:12]: CPU limits aren't set on Pods, they're set on containers.

File: other/require-cpu-limits/.chainsaw-test/pod-bad.yaml
Comment (chipzoller) [inline other/require-cpu-limits/.chainsaw-test/pod-bad.yaml:1]: This test resource manifest needs to cover cases where there are multiple containers (both `containers[]` and `initContainers[]`) in which one container sets a limit while the other does not. You also do not need to include any securityContext fields as these are irrelevant to this policy.

File: other/require-cpu-limits/.kyverno-test/resource.yaml
Comment (chipzoller) [inline other/require-cpu-limits/.kyverno-test/resource.yaml:1]: See similar comment on Chainsaw resource file.

**Review by chipzoller — CHANGES_REQUESTED** [review]

**Comment by nsagark** [comment]
@chipzoller I have updated the files as per your comments. Please review

**Comment by chipzoller** [comment]
Are all comments resolved?

**Comment by nsagark** [comment]
Hi @chipzoller Yes, I have updated the files. Just to make sure to address below comment,

`This test resource manifest needs to cover cases where there are multiple containers (both containers[] and initContainers[]) in which one container sets a limit while the other does not`

I have added below in resource.yaml file. Is that what you meant or something else?

`apiVersion: v1
kind: Pod
metadata:
  name: badpod07
spec:
  initContainers:
  - name: initcontainer01
    image: dummyimagename
  - name: initcontainer02
    image: dummyimagename
  containers:
  - name: container01
    image: dummyimagename
    resources:
      limits:
        cpu: "500m"
  - name: container02
    image: dummyimagename    `

**Comment by chipzoller** [comment]
Yes

**Comment by nsagark** [comment]
With that, I have addressed all the comments. Please review.

File: other/require-cpu-limits/.chainsaw-test/pod-bad.yaml [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+apiVersion: v1
+kind: Pod
+metadata:
+  name: badpod06
+spec:
+  initContainers:
+  - name: initcontainer01
+    image: busybox:1.35
+  - name: initcontainer02
+    image: busybox:1.35
```
Comment (chipzoller) [inline other/require-cpu-limits/.chainsaw-test/pod-bad.yaml:62]: Please update this initcontainer with a CPU limit.

File: other/require-cpu-limits/.chainsaw-test/pod-bad.yaml [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+---
+apiVersion: v1
+kind: Pod
+metadata:
+  name: badpod02
+spec:
+  containers:
+  - name: container01
+    image: busybox:1.35
```
Comment (chipzoller) [inline other/require-cpu-limits/.chainsaw-test/pod-bad.yaml:17]: This is redundant from badpod01. Remove.

File: other/require-cpu-limits/.chainsaw-test/pod-bad.yaml [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+apiVersion: v1
+kind: Pod
+metadata:
+  name: badpod04
+spec:
+  containers:
+  - name: container01
+    image: busybox:1.35
+  - name: container02
+    image: busybox:1.35
```
Comment (chipzoller) [inline other/require-cpu-limits/.chainsaw-test/pod-bad.yaml:39]: This is redundant from badpod03. Remove.

File: other/require-cpu-limits/.chainsaw-test/podcontroller-bad.yaml
Comment (chipzoller) [inline other/require-cpu-limits/.chainsaw-test/podcontroller-bad.yaml:1]: Please make similar changes to this file as those pointed out in `pod-bad.yaml`.

**Review by chipzoller — CHANGES_REQUESTED** [review]
Please ensure to resolve all conversations when complete.

**Comment by nsagark** [comment]
@chipzoller All conversations are resolved.

**Comment by chipzoller** [comment]
Several conversations unresolved.

**Comment by nsagark** [comment]
@chipzoller Sorry, I have resolved previous conversations. Please check now

**Review by chipzoller — CHANGES_REQUESTED** [review]
Looks good other than the AH digest needing to be updated.

**Comment by nsagark** [comment]
Hi @chipzoller The AH digest seems to be correct. Am I doing anything wrong below?

Sagar@DESKTOP-VS85098 MINGW64 /c/users/sagar/Downloads/kyverno-prs/policies/other/require-cpu-limits (require-cpu-limits)
$ sha256sum require-cpu-limits.yaml
2fe9fa043fbb4c85da0204a00a60ce6732ac0dc0e3a5ece4f7b741982a1003d1 *require-cpu-limits.yaml

Sagar@DESKTOP-VS85098 MINGW64 /c/users/sagar/Downloads/kyverno-prs/policies/other/require-cpu-limits (require-cpu-limits)
$ ls -lrht
total 8.0K
-rw-r--r-- 1 Sagar 197121 1.5K Jul 31 09:47 require-cpu-limits.yaml
-rw-r--r-- 1 Sagar 197121 1.5K Jul 31 09:47 artifacthub-pkg.yml

**Comment by chipzoller** [comment]
You're not doing anything wrong, it's just the expected digest is calculated in the pipeline under Linux. You're doing it under Windows which means the digest may be slightly different. The one being reported in the pipeline is `1d9998010342080ae64f309befdf52065de557cdcc10ddf68d8476b5af93d505`. Go ahead and put that in the AH metadata file and it should pass.

**Comment by nsagark** [comment]
Hi @chipzoller I have updated the digest. Thanks for pointing it out.

**Comment by nsagark** [comment]
Hi @chipzoller I don't see the changes yet in the main branch. Are we good here or anything missing?

**Review by chipzoller — APPROVED** [review]

**Pull request merged**